File: bench/vec.rb

class Vec2D
	def initialize(x, y)
		@x = x
		@y = y
	end
	
	attr_accessor :x
	attr_accessor :y
	
	def +(v)
		Vec2D.new(@x+v.x, @y+v.y)
	end
	
	def add_assign(v)
		@x += v.x
		@y += v.y
		return self
	end
	
	def to_s
		"Vec2D.new(x=#{@x}, y=#{@y})"
	end
end


def foo()
	v = Vec2D.new(0, 0)
	u = Vec2D.new(1, 2)
	i = 0
	while i<5000000
		#v = v + u
		v = v.add_assign(u)
		i += 1
	end
	
	puts v
end

foo

GC.start

File: bench/vec.lua

Vec2D = {}

function Vec2D.new(a, b)
    local tmp = {
        x = a,
        y = b
    }
    return setmetatable(tmp, { __index = Vec2D })
end

function Vec2D:add_assign(v)
    self.x = self.x + v.x
    self.y = self.y + v.y
    return self
end

function Vec2D:to_s()
    return "Vec2D.new(" .. self.x .. ", " .. self.y .. ")"
end

function foo()
	local v = Vec2D.new(0, 0)
	local u = Vec2D.new(1, 2)
	local i = 0
	while i<5000000 do
		v = v:add_assign(u)
		i = i + 1
	end
	
	print(v:to_s())
end

foo()
